mesonbuild/compilers/clike.py

```diff
@@ -306,7 +307,8 @@ def sanity_check_impl(self, work_dir, environment, sname, code):
                 # a ton of compiler flags to differentiate between
                 # arm and x86_64. So just compile.
                 mode = 'compile'
```

Comment:
I think this mode line is suppose to already handle that, so the bug is elsewhere.

Reply:
`mode` here is only passed to `self._get_basic_compiler_args()`.

`self._get_basic_compiler_args()` is used both here (in `sanity_check_impl()`) and in `self._get_compiler_check_args()` is called a number of places, including some that are passing these args to `self.compile()`, for example: `self._build_wrapper()`. `self._build_wrapper()` has it's own `mode` argument which it propagates into `self.compile()`, and `self.compile()` calls `self.get_compile_only_args()` itself.

The result is that we don't want to modify `self._get_basic_compiler_args()` to include `self.get_compile_only_args()` because it would result in duplication of `self.get_compile_only_args()`.

`sanity_check_impl()` notably does not use `self.compile()`, so it doesn't get the `self.get_compile_only_args()` added.

It's plausible there should be some other restructuring so that `sanity_check_impl()` has something to ask for args that understand's `mode`, but it's somewhat unclear what that should be, and probably would need larger numbers of changes.

Reply:
@jmesmon thanks for getting to the bottom of this! I'm not sure if you saw #5496, but this issue affects sanity checking and feature checking alike. Thanks for diagnosing that for me :) Furthermore, last I checked the `is_cross` / `exe_wrapper` logic was duplicated across the languages. (I assume @dcbaker's changes couldn't have totally fixed this as the non-C-family ones are affected too.)

I leave it to you to decide what refactor is best. Modifying `compile` to not include `_get_basic_compiler_args` as you say sounds good. Maybe `_get_basic_compiler_args` can even be inlined away!

> It's plausible there should be some other restructuring so that sanity_check_impl() has something to ask for args that understand's mode

That also sounds good!

That all said, don't feel pressured to fix all these things, just doing a first step and laying out a roadmap for later refactors is also much appreciated.

Reply:
> I'm not sure if you saw #5496, but this issue affects sanity checking and feature checking alike
I saw #5496, but I had not noticed that it affected feature checking.


After looking into this a bit: modifying `Compiler.compile` to drop the use of `self.get_compile_only_args()` would require that all callers now pass their `self.get_compile_only_args()` in `extra_args`. This seems not great (as `Compiler.compile` takes a mode anyhow), so it seems like factoring out the arg generation so that other methods can use it would make sense. `CLikeCompiler.sanity_check_impl`, would then use that function instead of using their `self._get_basic_compile_args()` directly. But the flags generated for the mode in `Compiler.compile` seem somewhat different from those currently generated in `CLikeCompiler._get_basic_compile_args`, making it unclear to me if they're related enough for re-use to occur. `CLikeCompiler._get_basic_compile_args()` omits `Compiler.get_always_args()`.

If we're ok with passing `Compiler.get_always_args()` in the `CLikeCompiler.sanity_check_impl()`, then we can probably reuse the flag generation from `Compiler.compile` in `CLikeCompiler.sanity_check_impl()`.

I've dug around a bunch an written an incomplete graphviz dot file to try to get an understanding of the flow of compiler arguments, but it's pretty unclear what's going on.

<details><summary>incomplete-arg-flow.dot</summary>

```
digraph {
	graph [rankdir=LR];
	compiler___get_linker_debug_crt_args -> compiler___get_basic_compiler_args;
	compiler___get_compile_only_args -> compiler___get_basic_compiler_args;
	env_coredata_get_external_args -> compiler___get_basic_compiler_args;
	env_coredata_get_external_link_args -> compiler___get_basic_compiler_args;

	compiler___get_basic_compiler_args -> ClikeCompiler___get_compiler_check_args;
	dep__get_compile_args -> ClikeCompiler___get_compiler_check_args;
	dep__get_link_args -> ClikeCompiler___get_compiler_check_args;
	compiler__get_compiler_check_args -> ClikeCompiler___get_compiler_check_args;

	compiler__get_compile_only_args -> ninjabackend__generate_compile_rule_for;
	ninjabackend__generate_compile_rule_for -> ninjabackend__generate_compile_rules;
	ninjabackend__generate_compile_rules -> ninjabackend__generate_rules;
	ninjabackend__generate_link -> ninjabackend__generate_static_link_rules;
	ninjabackend__generate_link -> ninjabackend__generate_dynamic_link_rules;
	ninjabackend__generate_static_link_rules -> ninjabackend__generate_rules;
	ninjabackend__generate_dynamic_link_rules -> ninjabackend__generate_rules;

	compiler__get_compile_only_args -> Compiler__compile;

	compiler__get_always_args -> Compiler__compile;
	compiler__get_preprocess_only_args -> Compiler__compile;
	compiler__get_output_args -> Compiler__compile;

	compiler__get_always_args -> ClikeCompiler__find_framework_paths;
	env_coredata_get_external_args -> ClikeCompiler__find_framework_paths;

	linker__get_linker_always_args -> ninjabackend__generate_link;
	compilers__get_base_link_args -> ninjabackend__generate_link;
	ninjabackend__get_no_stdlib_link_args -> ninjabackend__get_cross_stdlib_link_args
	ninjabackend__get_cross_stdlib_link_args -> ninjabackend__generate_link;
	ninjabackend__get_link_whole_args -> ninjabackend__generate_link;
	ninjabackend__get_link_debugfile_args -> ninjabackend__generate_link;
	ninjabackend__get_target_type_link_args -> ninjabackend__generate_link;
	ninjabackend__build_target_link_arguments -> ninjabackend__generate_link;
	ninjabackend__get_target_type_link_args_post_dependencies -> ninjabackend__generate_link;
	linker__get_option_link_args -> ninjabackend__generate_link;
	linker__build_rpath_args -> ninjabackend__generate_link;

	backend__get_compiler_options_for_target -> backend__generate_basic_compiler_args;
	backend__get_cross_stdlib_args -> backend__generate_basic_compiler_args;
	compiler__get_always_args -> backend__generate_basic_compiler_args;
	compiler__get_no_warn_args -> backend__generate_basic_compiler_args;
	compiler__get_warn_args -> backend__generate_basic_compiler_args;
	compiler__get_werror_args -> backend__generate_basic_compiler_args;
	compiler__get_option_compile_args -> backend__generate_basic_compiler_args;
	compiler__get_optimization_args -> backend__generate_basic_compiler_args;
	compiler__get_debug_args -> backend__generate_basic_compiler_args;
	backend_build__get_project_args -> backend__generate_basic_compiler_args;
	backend_build__get_global_args -> backend__generate_basic_compiler_args;
	env_coredata_get_external_args -> backend__generate_basic_compiler_args;
	compiler__get_pic_args -> backend__generate_basic_compiler_args;
	compiler__get_pie_args -> backend__generate_basic_compiler_args;
	dep__get_compile_args -> backend__generate_basic_compiler_args;
	dep__get_exe_args -> backend__generate_basic_compiler_args;

	compiler__gnu_symbol_visibility_args -> ninjabackend___generate_single_compile;
	compilers__get_base_compile_args -> ninjabackend___generate_single_compile;
	backend__generate_basic_compiler_args -> ninjabackend___generate_single_compile;
	ninjabackend__get_custom_target_dir_include_args -> ninjabackend___generate_single_compile;
	// for each dep & for get_target_private_dir()
	compiler__get_include_args -> ninjabackend___generate_single_compile;
	ninjabackend__get_source_dir_include_args -> ninjabackend___generate_single_compile;
	ninjabackend__get_build_dir_include_args -> ninjabackend___generate_single_compile;

	linker_get_buildtype_linker_args -> ninjabackend__generate_link;
	env_coredata_get_external_link_args -> ninjabackend__generate_link;
	dep__get_link_args -> ninjabackend__generate_link;
	ClikeCompiler___get_compiler_check_args -> ClikeCompiler___build_wrapper;
	ClikeCompiler___build_wrapper -> ClikeCompiler__compiles;
	ClikeCompiler___build_wrapper -> ClikeCompiler__run;
	ClikeCompiler__compiles -> ClikeCompiler__links;

	// get_option_link_args()
	// get_option_compile_args()
	// get_pch_use_args()

	// get_colorout_args()
	// get_warn_args()
	// openmp_flags()
}
```
</details>

Reply:
@jmesmon thanks for more investigation. I wouldn't worry too much about preserving behavior, since the current state has regressed. Do the simplest thing that makes sense to you, history be damned, then we'll look at the diff.